### PR TITLE
Add documentation to many core editor types

### PR DESCRIPTION
**Pull request by vultix**
Hopefully this makes it a bit easier for new contributors to dive into the codebase :)

Release Notes:

- Improved documentation for many core editor types

### CODE REVIEW

File: .cargo/config.toml [hunk 1]
```diff
@@ -1,3 +1,7 @@
 [build]
 # v0 mangling scheme provides more detailed backtraces around closures
 rustflags = ["-C", "symbol-mangling-version=v0"]
+
+# Shows private items in the generated documentation when running `cargo doc`
+# Makes rustdoc more useful for understanding the codebase, not just the public api surface
+rustdocflags = ["--document-private-items"]
```
Comment (vultix) [inline .cargo/config.toml:7]: I'm not positive we want this flag here.

I added it because there are many items that are in private modules that have documentation that's useful for understanding the codebase. 

An example of this would be the module documentation for `editor::display_map`.  The existing documentation referenced many private types, such as `InlayMap`, making `cargo doc` not useful for understanding this part of the codebase.

Comment (vultix) [reply .cargo/config.toml:7]: I think the question we should ask is "will we publish these crates to crates.io and docs.rs?"  

If the answer is yes, we should disable this flag, and adjust our documentation to only reference public items.

If the answer is no, then `cargo doc` is an incredibly useful tool for understanding the local codebase, and I feel like this change is warranted.

File: crates/editor/src/display_map.rs [hunk 2]
```diff
@@ -67,15 +67,27 @@ pub trait ToDisplayPoint {
...
+/// folding, hard tabs, soft wrapping, custom blocks (like diagnostics), and highlighting.
+///
+/// See the [module level documentation](self) for more information.
 pub struct DisplayMap {
+    /// The buffer that we are displaying.
     buffer: Model<MultiBuffer>,
     buffer_subscription: BufferSubscription,
-    fold_map: FoldMap,
+    /// Decides where the [`Inlay`]s should be displayed.
     inlay_map: InlayMap,
```
Comment (vultix) [inline crates/editor/src/display_map.rs:79]: This is the only structural change I made in this pull request, and I'm happy to undo this change if we don't want to include it.

I moved `inlay_map` above `fold_map` to better represent the hierarchy of how these maps are applied.

File: crates/sum_tree/src/sum_tree.rs [hunk 3]
```diff
@@ -113,6 +151,10 @@ impl Bias {
     }
 }
 
+/// A binary tree where each leaf node contains an [`Item`] of type `T`,
```
Comment (mikayla-maki) [inline crates/sum_tree/src/sum_tree.rs:154]: This is incorrect, a SumTree is a B-Tree, not a binary tree

**Review by mikayla-maki — CHANGES_REQUESTED** [review]

**Comment by mikayla-maki** [comment]
Thank you for writing some documentation!

File: .cargo/config.toml (hunk 1 above)
Comment (mikayla-maki) [reply .cargo/config.toml:7]: I don't think we're planning on publishing these docs anywhere, they're entirely internal and won't be exposed otherwise.

File: crates/sum_tree/src/sum_tree.rs (hunk 3 above)
Comment (vultix) [reply crates/sum_tree/src/sum_tree.rs]: Good catch, thank you!

File: .cargo/config.toml (hunk 1 above)
Comment (mikayla-maki) [reply .cargo/config.toml:7]: We're probably not going to be publishing these as docs anytime soon. That said, let's let folks pick which documentation they prefer. This should be removed.

Comment (maxdeviant) [reply .cargo/config.toml:7]: I don't think we should be including private items in rustdoc, as it makes it less clear what the public interface of a given crate is.

**Review by mikayla-maki — APPROVED** [review]

File: .cargo/config.toml [hunk 4]
```diff
@@ -1,3 +1,7 @@
 [build]
 # v0 mangling scheme provides more detailed backtraces around closures
 rustflags = ["-C", "symbol-mangling-version=v0"]
+
+# Shows private items in the generated documentation when running `cargo doc`
+# Makes rustdoc more useful for understanding the codebase, not just the public api surface
```
Comment (mikayla-maki) [inline .cargo/config.toml:6]: This needs to be removed.

**Review by mikayla-maki — CHANGES_REQUESTED** [review]

File: .cargo/config.toml (hunk 1 above)
Comment (vultix) [reply .cargo/config.toml]: Makes sense, I've removed the flag!

**Review by mikayla-maki — APPROVED** [review]

**Comment by vultix** [comment]
Not sure about the CI failure, is there something I need to do there?

**Comment by nathansobo** [comment]
Looks like it was a couple of warnings. I just pushed a commit that hopefully addresses them.

**Comment by nathansobo** [comment]
Thanks!
